### ISSUE TITLE
GDScript 2.0: Setting a typed variable to a previously freed instance causes a freeze and crash without report to console

### ISSUE THREAD

**Issue by wickworks** [godotengine/godot#71771]
### Godot version

4.0.beta14.mono.official

### System information

macOS Monterey 12.5.1

### Issue description

Trying to set a typed variable to a previously freed instance causes the game to freeze and then crash without any debug information being sent to the console.

I would expect it to fail and either continue (so you could for example proceed to call `is_instance_valid` on it) or throw an error, but it does neither.

### Steps to reproduce

When the object passed in as `instance` has been freed, this code works fine:
```
func print_instance_untyped(instance):
	var user_class = instance
	print(user_class)
```

But this code causes a silent crash:
```
func print_instance_typed(instance):
	var user_class:CustomGameClass = instance
	print(user_class)
```

(It doesn't matter if the type is `CustomGameClass` or a built-in type; e.g. it crashes the same when it's a `Label`.

### Minimal reproduction project

[FreedTypedVariableCrash.zip](https://github.com/godotengine/godot/files/10470921/FreedTypedVariableCrash.zip)

**Comment by anvilfolk** [comment]
I'm starting to try to take a look but have very very little capacity right now, so if someone hops in & fixes it that's really cool too :)

For right now I just confirmed the issue on master (post RC2), and got the stack trace:

```
================================================================
CrashHandlerException: Program crashed
Engine version: Godot Engine v4.0.rc.custom_build (830eff64a75a9cad211a60e91a2d804592d4d7e8)
Dumping the backtrace. Please include this when reporting the bug to the project developer.
[0] Object::get_class_name (C:\dev\godot\core\object\object.h:794)
[1] GDScriptFunction::call (C:\dev\godot\modules\gdscript\gdscript_vm.cpp:1280)
[2] GDScriptInstance::callp (C:\dev\godot\modules\gdscript\gdscript.cpp:1826)
[3] Object::callp (C:\dev\godot\core\object\object.cpp:711)
[4] Variant::callp (C:\dev\godot\core\variant\variant_call.cpp:1161)
[5] GDScriptFunction::call (C:\dev\godot\modules\gdscript\gdscript_vm.cpp:1613)
[6] GDScriptInstance::callp (C:\dev\godot\modules\gdscript\gdscript.cpp:1826)
[7] Node::_gdvirtual__input_call<0> (C:\dev\godot\scene\main\node.h:245)
[8] Node::_call_input (C:\dev\godot\scene\main\node.cpp:2768)
[9] SceneTree::_call_input_pause (C:\dev\godot\scene\main\scene_tree.cpp:924)
[10] Viewport::push_input (C:\dev\godot\scene\main\viewport.cpp:2833)
[11] Window::_window_input (C:\dev\godot\scene\main\window.cpp:1373)
[12] call_with_variant_args_helper<Window,Ref<InputEvent> const &,0> (C:\dev\godot\core\variant\binder_common.h:293)
[13] call_with_variant_args<Window,Ref<InputEvent> const &> (C:\dev\godot\core\variant\binder_common.h:408)
[14] CallableCustomMethodPointer<Window,Ref<InputEvent> const &>::call (C:\dev\godot\core\object\callable_method_pointe)
[15] Callable::callp (C:\dev\godot\core\variant\callable.cpp:51)
[16] DisplayServerWindows::_dispatch_input_event (C:\dev\godot\platform\windows\display_server_windows.cpp:2314)
[17] DisplayServerWindows::_dispatch_input_events (C:\dev\godot\platform\windows\display_server_windows.cpp:2279)
[18] Input::_parse_input_event_impl (C:\dev\godot\core\input\input.cpp:690)
[19] Input::flush_buffered_events (C:\dev\godot\core\input\input.cpp:941)
[20] DisplayServerWindows::process_events (C:\dev\godot\platform\windows\display_server_windows.cpp:2008)
[21] OS_Windows::run (C:\dev\godot\platform\windows\os_windows.cpp:1296)
[22] widechar_main (C:\dev\godot\platform\windows\godot_windows.cpp:181)
[23] _main (C:\dev\godot\platform\windows\godot_windows.cpp:203)
[24] main (C:\dev\godot\platform\windows\godot_windows.cpp:217)
[25] WinMain (C:\dev\godot\platform\windows\godot_windows.cpp:231)
[26] __scrt_common_main_seh (D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288)
[27] <couldn't map PC to fn name>
-- END OF BACKTRACE --
================================================================
```

**Comment by anvilfolk** [comment]
Made a fix in #73364, but need some feedback on what we want the behavior to be, so please head on over & share your thoughts!
